File: practices/01-golden-workflow/src/todo.py
"""Todo management app - Practice 01 hands-on code"""

from dataclasses import dataclass, field
from datetime import datetime
from typing import Optional


@dataclass
class Todo:
    id: int
    title: str
    completed: bool = False
    created_at: datetime = field(default_factory=datetime.now)


_todos: list[Todo] = []
_next_id: int = 1


def add_todo(title: str) -> Todo:
    global _next_id
    todo = Todo(id=_next_id, title=title)
    _next_id += 1
    _todos.append(todo)
    return todo


def complete_todo(id: int) -> Optional[Todo]:
    for todo in _todos:
        if todo.id == id:
            todo.completed = True
            return todo
    return None


def delete_todo(id: int) -> bool:
    for i, todo in enumerate(_todos):
        if todo.id == id:
            _todos.pop(i)
            return True
    return False


def list_todos() -> list[Todo]:
    return list(_todos)


def list_completed() -> list[Todo]:
    return [t for t in _todos if t.completed]


def list_pending() -> list[Todo]:
    return [t for t in _todos if not t.completed]


def clear_all() -> None:
    global _next_id
    _todos.clear()
    _next_id = 1

File: practices/01-golden-workflow/src/todo.ts
// Todo 관리 앱 - Practice 01 실습용 코드

export interface Todo {
  id: number;
  title: string;
  completed: boolean;
  createdAt: Date;
}

let todos: Todo[] = [];
let nextId = 1;

export function addTodo(title: string): Todo {
  const todo: Todo = {
    id: nextId++,
    title,
    completed: false,
    createdAt: new Date(),
  };
  todos.push(todo);
  return todo;
}

export function completeTodo(id: number): Todo | undefined {
  const todo = todos.find((t) => t.id === id);
  if (todo) {
    todo.completed = true;
  }
  return todo;
}

export function deleteTodo(id: number): boolean {
  const index = todos.findIndex((t) => t.id === id);
  if (index !== -1) {
    todos.splice(index, 1);
    return true;
  }
  return false;
}

export function listTodos(): Todo[] {
  return [...todos];
}

export function listCompleted(): Todo[] {
  return todos.filter((t) => t.completed);
}

export function listPending(): Todo[] {
  return todos.filter((t) => !t.completed);
}

export function clearAll(): void {
  todos = [];
  nextId = 1;
}
